src/java/org/apache/cassandra/index/sai/disk/v1/segment/SegmentRamBuffer.java

```diff
@@ -73,9 +73,9 @@ public long add(int segmentRowId, byte[] value)
         return (trie.sizeOnHeap() - initialSizeOnHeap) + (postingsAccumulator.heapAllocations() - reducerHeapSize);
     }
 
-    public BlockBalancedTreeIterator iterator()
+    public SegmentRamIterator iterator()
     {
-        return BlockBalancedTreeIterator.fromTrieIterator(trie.entrySet().iterator(), bytesPerValue);
+        return SegmentRamIterator.fromTrieIterator(trie.entrySet().iterator());
```

Comment:
What if we inlined `fromTrieIterator()` here and moved `fromTermsIterator()` to `TermsIterator`, and then just got rid of the `SegmentRamIterator` in favor of `Iterator<Pair<ByteComparable, PostingList>>`? `SegmentRamBuffer` isn't bloated by any means, and we could even make it `Iterable` in that case, given it already provides `iterator()`. `SegmentRamIterator` is really just half "marker" interface and half utility class for static factories.

If `Iterator<Pair<ByteComparable, PostingList>>` feels to complicated, the solution there has probably aways been to stop using `Pair` and have something like `IndexEntry` or `TermAndPostings` (or something else that communicates well).

Reply:
I agree that the `Pair` usage has got out of hand with the SAI code, so we can start by getting rid of it here. 

I will have a look at the iterator transformations as well. This has been through a number of iterations during this ticket.

Reply:
I have removed `SegmentRamIterator` and made `TermsIterator` an `Interator<IndexEntry>`. This has simplified things a lot. `SegmentRamBuffer` is now `SegmentTrieBuffer`. `IndexEntry` is preferred because that is used in the internal index to represent the same thing. So we are keeping common terminology.